include/download_client.h

```diff
@@ -17,220 +17,184 @@
  *
- * @param[in] client	The client instance.
- * @param[in] event     The event.
- * @param[in] status    Event status (either 0 or an errno value).
+ * If the callback returns a non-zero value, the download will stop.
+ * To resume the download, use @ref download_client_start().
  *
- * @retval 0 If the event was handled successfully.
- *           Other values indicate that the application failed to handle
- *           the event.
+ * @param[in] event	The event.
+ *
+ * @return Zero to continue the download, non-zero otherwise.
  */
-typedef int (*download_client_event_handler_t)(struct download_client *client,
-				enum download_client_evt event, u32_t status);
+typedef int (*download_client_callback_t)(
+	const struct download_client_evt *event);
 
-/** @brief Object download client instance that describes the state of download.
+/**
+ * @brief Download client instance.
  */
 struct download_client {
-	/** Buffer used to receive responses from the server.
-	 *  This buffer can be read by the application if necessary,
-	 *  but must never be written to.
-	 */
-	char resp_buf[CONFIG_NRF_DOWNLOAD_MAX_RESPONSE_SIZE];
-	/** Buffer used to create requests to the server.
-	 *  This buffer can be read by the application if necessary,
-	 *  but must never be written to.
-	 */
-	char req_buf[CONFIG_NRF_DOWNLOAD_MAX_REQUEST_SIZE];
-	/** Pointer to the object fragment in @p resp_buf.
-	 *  The response from the server contains protocol metadata
-	 *  in addition to the object fragment. On every
-	 *  DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG event, this pointer is
-	 *  updated to point to the latest object fragment.
-	 *  This pointer shall not be updated by the application.
-	 */
-	char *fragment;
-	/** Size of the fragment. The size is updated on every
-	 *  DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG event.
-	 */
-	int fragment_size;
-	/** Transport file descriptor.
-	 *  If negative, the transport is disconnected.
-	 */
+	/** Transport file descriptor */
 	int fd;
-	/** Total size of the object being downloaded.
-	 * If negative, the download is in progress.
-	 * If zero, the size is unknown.
+	/** HTTP response buffer */
+	char buf[CONFIG_NRF_DOWNLOAD_MAX_RESPONSE_SIZE];
+	/** Buffer offset */
+	size_t offset;
+
+	/** Size of the file being downloaded, in bytes */
+	size_t file_size;
+	/** Download progress, number of bytes downloaded */
+	size_t progress;
+	/** Whether the HTTP header for
+	 * the current fragment has been processed
 	 */
-	int object_size;
-	/** Current size of the object being downloaded. */
-	volatile int download_size;
-	/** Status of the transfer (see @ref download_client_status). */
-	volatile int status;
-	/** Server that hosts the object. */
-	const char *host;
-	/** Resource to be downloaded. */
-	const char *resource;
-	/** Event handler. Must not be NULL. */
-	const download_client_event_handler_t callback;
+	bool has_header;
+
+	/** Server hosting the file, NULL terminated */
+	char *host;
```

Comment:
Is there any reason why this and `file` is not `const` ?

Reply:
That would limit one instance of the library to one server and one file, but in practice that limitation is artificial since the implementation can connect to a new server and download a new file. By letting both variables be modifiable, you can re-use one instance to download different files from different servers.